src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala

```diff
@@ -71,26 +72,79 @@ class CirceScroogeMacrosImpl(val c: blackbox.Context) {
+            * If the field is present but fails to decode then it's still an error.
+            * Note: reverse-engineering the name of the default value because it's not easily available in the reflection API.
+            */
+          val defaultValue = A.companion.member(TermName("apply$default$" + (i + 1)))
+          fq"""$fresh <- $decodeParam.getOrElse(_root_.scala.Right($defaultValue))"""
+        } else {
+          fq"""$fresh <- $decodeParam.getOrElse(_root_.scala.Left(_root_.io.circe.DecodingFailure("Missing field: " + ${name.toString}, Nil)))"""
+        }
+      }
+
+      val accDecodeExpr = {
+        /**
+          * If we try to use the same tree twice, the compiler may blow up with the following helpful error message:
+          *   `scala.reflect.internal.Types$TypeError: value <none> is not a member of com.gu.fezziwig.FezziwigTests`
+          */
+        val decoderCopy = c.untypecheck(implicitDecoder)
+
+        val accDecodeParam =
+          q"""cursor.downField(${name.toString}).success
+            .map(x => $decoderCopy.accumulating(x))"""
 
-      val expr =
         if (param.asTerm.isParamWithDefault) {
-          // Fallback to param's default value if the JSON field is not present (or can't be decoded for some other reason).
-          // Note: reverse-engineering the name of the default value because it's not easily available in the reflection API.
           val defaultValue = A.companion.member(TermName("apply$default$" + (i + 1)))
-          fq"""$fresh <- $decodeParam.orElse(_root_.scala.Right($defaultValue))"""
-        } else
-          fq"""$fresh <- $decodeParam"""
-      (fresh, expr)
+          q"""$accDecodeParam.getOrElse(_root_.cats.data.Validated.valid[_root_.cats.data.NonEmptyList[_root_.io.circe.DecodingFailure],$tpe]($defaultValue))"""
+        } else {
+          q"""$accDecodeParam.getOrElse(_root_.cats.data.Validated.invalidNel(_root_.io.circe.DecodingFailure("Missing field: " + ${name.toString}, cursor.history)))"""
+        }
+      }
+
+      (fresh, decodeExpr, accDecodeExpr)
+    }
+
+    /**
+      * Accumulate cats Validated results using cartesian |@| operator, e.g.
+      * (expr1 |@| expr2 |@| ...) map (apply)
+      *
+      * This is only supported for structs with <= 22 parameters, because of a limitation in cats.
+      */
+    val accumulating: Option[Tree] = {
+      if (params.length <= 22) {
+        val validation = params.map(_._3).reduce { (acc: Tree, expr: Tree) =>
+          q"""$acc.|@|($expr)"""
+        }
+        Some(q"""
+          override def decodeAccumulating(cursor: _root_.io.circe.HCursor): _root_.io.circe.AccumulatingDecoder.Result[$A] = {
+            ($validation) map ($apply)
+          }
+        """)
+      } else {
+        c.warning(c.enclosingPosition, s"Decoder for ThriftStruct ${A.typeSymbol.name.toString} will not support accumulated errors for nested types because it has more than 22 parameters.")
```

Comment:
Not that this relates to this PR, but Shapeless offers a way around this with HList. Not sure if this could help if this ever became a problem?